man/dmfs.1
.TH "DMFS" "1" "" "Version 2.0.0" "User Commands"
.SH NAME
dmfs \(en reads observations from file or named pipe
.SH SYNOPSIS
\f[B]dmfs\f[R] \-\-help
.PP
\f[B]dmfs\f[R] \-\-version
.PP
\f[B]dmfs\f[R] \-\-\f[B]config\f[R] \f[I]file\f[R] [\-\-\f[B]name\f[R]
\f[I]name\f[R]] [\-\-\f[B]node\f[R] \f[I]id\f[R]] [\-\-\f[B]sensor\f[R]
\f[I]id\f[R]] [\-\-\f[B]logger\f[R] \f[I]name\f[R]]
[\-\-\f[B]format\f[R] \f[I]format\f[R]] [\-\-\f[B]output\f[R]
\f[I]file\f[R]] [\-\-\f[B]debug\f[R]]
.SH DESCRIPTION
The \f[B]dmfs\f[R] program reads observations from file system, virtual
file, or named pipe.
The program can be used to read sensor values from the 1\-Wire File
System (OWFS).
.PP
All requests of an observation have to contain the path of the file to
be read in attribute \f[CR]request\f[R].
Response values are extracted by group from the raw response using the
given regular expression pattern.
For each group to extract, a response with the name of the respective
group has to be added to the request.
.PP
If any receivers are specified, observations are forwarded to the next
receiver via POSIX message queue.
The program can act as a sole data logger if output and format are set.
If the output path is set to \f[CR]\-\f[R], observations are printed to
\f[I]stdout\f[R].
.PP
A configuration file is required to configure the jobs to perform.
Each observation must have a valid target id.
The database must contain the node, the sensor, and the target.
.SH OPTIONS
.TP
\f[B]\-\-config\f[R], \f[B]\-c\f[R] \f[I]file\f[R]
File path to the configuration file.
.TP
\f[B]\-\-debug\f[R], \f[B]\-D\f[R]
Forward logs messages of level \f[CR]LL_DEBUG\f[R] via IPC (if logger is
set).
.TP
\f[B]\-\-format\f[R], \f[B]\-f\f[R] [csv|jsonl]
Output format of observations if \f[CR]\-\-output\f[R] is set.
Either CSV or JSON Lines.
.TP
\f[B]\-\-help\f[R], \f[B]\-h\f[R]
Print available command\-line arguments and quit.
.TP
\f[B]\-\-logger\f[R], \f[B]\-l\f[R] \f[I]name\f[R]
Name of logger.
If set, sends logs to \f[I]dmlogger(1)\f[R] process of given name.
.TP
\f[B]\-\-name\f[R], \f[B]\-n\f[R] \f[I]name\f[R]
Name of program instance and configuration (default is \f[CR]dmfs\f[R]).
.TP
\f[B]\-\-node\f[R], \f[B]\-N\f[R] \f[I]id\f[R]
Node id.
.TP
\f[B]\-\-output\f[R], \f[B]\-o\f[R] \f[I]file\f[R]
Output file to append observations to (\f[CR]\-\f[R] for
\f[I]stdout\f[R]).
.TP
\f[B]\-\-sensor\f[R], \f[B]\-S\f[R] \f[I]id\f[R]
Sensor id.
.TP
\f[B]\-\-verbose\f[R], \f[B]\-V\f[R]
Print log messages to \f[I]stderr\f[R].
.TP
\f[B]\-\-version\f[R], \f[B]\-v\f[R]
Print version information and quit.
.SH EXIT STATUS
.TP
\f[B]0\f[R]
Success.
Process terminated without errors.
.TP
\f[B]1\f[R]
Failure.
Process failed.
.SH ENVIRONMENT
.TP
\f[B]DM_LOGGER\f[R]
Name of logger instance to send logs to.
.TP
\f[B]NO_COLOR\f[R]
Disable ANSI colour output.
.SH EXAMPLE
Read the jobs to perform from configuration file and execute them
sequentially:
.IP
.EX
$ dmfs \-n dmfs \-c /opt/etc/dmpack/dmfs.conf \-V
.EE


CHANGELOG.md
# Change Log

All notable changes to the project will be documented in this file.

## [2.0.0] – Unreleased

## Library

* `Added` more error codes to `dm_error`
* `Added` more log levels to `dm_log`
* `Added` _fortran-zeromq_ interface bindings to `vendor/`
* `Added` function `dm_net_ipv6_is_valid()` to module `dm_net`
* `Added` type `posix_signal_type` and self-pipe procedures to module `dm_posix_signal`
* `Added` module `dm_arg_parser` (outsourced from `dm_arg`)
* `Added` module `dm_gantner` for eGateHighSpeedPort API access
* `Added` module `dm_group` for observation groups
* `Added` module `dm_job_list` (outsourced from `dm_job`)
* `Added` module `dm_msgpack` for MessagePack serialisation/deserialisation
* `Added` module `dm_random` for pseudo-random number generation
* `Added` module `dm_zmq` for ZeroMQ access
* `Changed` location of sleep routines (moved to `dm_posix`)
* `Changed` name of module `dm_sql` to `dm_db_sql`
* `Changed` name of module `dm_system` to `dm_posix`
* `Changed` names of modules `dm_fifo`, `dm_mqueue`, `dm_pipe`, `dm_sem`, `dm_signal`, `dm_thread`, `dm_tty` to include `posix` prefix
* `Changed` database schema due to new data model
* `Changed` default database suffix from `.sqlite` to `.db`
* `Changed` GeoCOM API for Lua to accept prototype observation as first argument
* `Changed` GeoCOM API response identifiers to more memorable names
* `Changed` function `dm_path_join` in `dm_path`
* `Changed` model of observation data structure in `dm_observ`, removed priority, requests, and receivers
* `Changed` model of sensor data structure in `dm_sensor`
* `Deleted` module `dm_mqueue_util`
* `Deleted` module `dm_request`
* `Fixed` GeoCOM API response identifiers

## Programs

* `Changed` response value units in `dmved`
* `Changed` signal handlers to use self-pipe for POSIX conformance

## Documentation

* Converted user guide from AsciiDoc to Markdown
* Converted man pages from AsciiDoc to Markdown

## [1.0.1] – 2026-04-28

### Library

* `Added` example filter program to `share/dmfilter/`

## [1.0.0] – Unreleased

## Library

* `Added` Bessel, Butterworth, Chebyshev filters to `dm_filter`

## [0.9.9] – 2026-01-11

### Library

* `Added` module `dm_coord` for coordinate transformation
* `Added` module `dm_filter` for filtering
* `Added` module `dm_ghostscript` for Ghostscript access
* `Added` module `dm_netstring` for netstring parsing
* `Added` module `dm_statistics` of statistics functions
* `Added` file tree size function to module `dm_file`
* `Added` file modification date/time to function `dm_file_touch()` in `dm_file`
* `Added` format `FORMAT_TSV` for tab-separated values to `dm_format`
* `Added` swap routines to `dm_util`
* `Changed` error code order for future additions
* `Changed` default database cache size of `dm_db_open()` to improve performance

### Programs

* `Added` author option to `dmreport`
* `Added` meta data to PDF output of `dmreport`
* `Fixed` reading of options in `dmdbctl`
* `Changed` last modified date/time to last log date/time in output of `dmfeed`

## [0.9.8] – 2025-08-30

### Library

* `Added` module `dm_image` for image handling
* `Added` module `dm_js` for JavaScript generation
* `Added` module `dm_roff` for GNU roff abstraction
* `Added` module `dm_serial` for derived type serialisation
* `Added` module `dm_transfer` for HTTP-RPC API file transfers
* `Added` GeoJSON Feature Collection procedures to `dm_geojson`
* `Added` image upload to `dm_rpc`
* `Added` HTTP request and response headers to `dm_rpc`
* `Added` HTTP response headers to `dm_fcgi`
* `Added` terminals `gpic`, `postscript`, and `sixeltek` to `dm_plot`
* `Added` image and transfer functions to `dm_db_api`
* `Changed` structure of database abstraction layer, added modules `dm_db_api`, `dm_db_count`, `dm_db_json`, `dm_db_pragma`, and `dm_db_row`
* `Changed` API of modules `dm_arg` and `dm_config` to object-oriented
* `Changed` query parameter API in `dm_cgi`
* `Changed` Git submodules to copies of third-party dependencies
* `Fixed` reading from pipe in `dm_pipe`
* `Fixed` unsigned type conversion in `dm_c`
* `Fixed` return code handling of database select functions

### Programs

* `Added` program `dmcamera` for capturing of camera images
* `Added` program `dmupload` for image upload to HTTP-RPC API
* `Added` scale factor for response values to `dmreport`
* `Added` PDF and PostScript output to `dmreport`
* `Added` message queue reading to `dmmb`
* `Added` image upload to `dmapi`
* `Added` image viewer to `dmweb`
* `Added` environment variables `DM_IMAGE_DB`, `DM_IMAGE_DIR` to `dmapi` and `dmweb`
* `Added` log file output to `dmlogger`
* `Changed` environment variables in `dmapi` and `dmweb` to `DM_BEAT_DB`, `DM_LOG_DB`, `DM_OBSERV_DB`
* `Changed` derived type serialisation in `dmapi` and `dmexport`
* `Changed` database access in `dmexport`
* `Changed` configuration of `dmgrc`

## [0.9.7] – 2025-04-05

### Library

* `Added` module `dm_c` for C interoperability
* `Added` module `dm_camera` for webcam and RTSP access (FFmpeg)
* `Added` module `dm_crypto` of cryptographic hash functions
* `Added` module `dm_db_query` for SQL query building
* `Added` module `dm_dwd` for DWD API format parsing
* `Added` module `dm_dwd_api` for DWD API access (libcurl)
* `Added` module `dm_ftp` for FTP transfer (libcurl)
* `Added` module `dm_freebsd` for system status access on FreeBSD
* `Added` module `dm_geojson` for GeoJSON serialisation
* `Added` module `dm_gm` for GraphicsMagick access
* `Added` module `dm_im` for XMPP connectivity (libstrophe)

app/dmfs.f90
! dmfs.f90
!
! Author:  Philipp Engel
! Licence: ISC
program dmfs
    !! Reads observations from file system (file, virtual file, or named pipe).
    use :: dmpack
    implicit none (type, external)

    character(*), parameter :: APP_NAME  = 'dmfs'
    integer,      parameter :: APP_MAJOR = 2
    integer,      parameter :: APP_MINOR = 0
    integer,      parameter :: APP_PATCH = 0

    character, parameter :: APP_CSV_SEPARATOR = ','    !! CSV field separator.
    logical,   parameter :: APP_MQ_BLOCKING   = .true. !! Observation forwarding is blocking.

    integer, parameter :: OUTPUT_NONE   = 0
    integer, parameter :: OUTPUT_STDOUT = 1
    integer, parameter :: OUTPUT_FILE   = 2

    type :: app_type
        !! Application settings.
        character(ID_LEN)          :: name        = APP_NAME    !! Instance and configuration name (required).
        character(FILE_PATH_LEN)   :: config      = ' '         !! Path to configuration file (required).
        character(LOGGER_NAME_LEN) :: logger      = ' '         !! Name of logger.
        character(NODE_ID_LEN)     :: node_id     = ' '         !! Node id (required).
        character(SENSOR_ID_LEN)   :: sensor_id   = ' '         !! Sensor id (required).
        character(FILE_PATH_LEN)   :: output      = ' '         !! Path of output file.
        character(FORMAT_NAME_LEN) :: format_name = 'none'      !! Output format name.
        integer                    :: output_type = OUTPUT_NONE !! Output type.
        integer                    :: format      = FORMAT_NONE !! Output format.
        logical                    :: debug       = .false.     !! Forward debug messages via IPC.
        logical                    :: verbose     = .false.     !! Print debug messages to stderr.
        type(job_list_type)        :: jobs                      !! Job list.
    end type app_type

    class(logger_class), pointer :: logger ! Logger object.

    integer                 :: rc     ! Return code.
    type(app_type)          :: app    ! App settings.
    type(posix_signal_type) :: signal ! Self-pipe.

    ! Initialise DMPACK.
    call dm_init()

    ! Get command-line arguments, read options from configuration file.
    rc = read_args(app)
    if (dm_is_error(rc)) call dm_stop(STOP_FAILURE)

    ! Initialise logger.
    logger => dm_logger_get_default()
    call logger%configure(name    = app%logger,  & ! Name of logger process.
                          node_id = app%node_id, & ! Node id.
                          source  = app%name,    & ! Log source.
                          debug   = app%debug,   & ! Forward debug messages via IPC.
                          ipc     = .true.,      & ! Enable IPC (if logger is set).
                          verbose = app%verbose)   ! Print logs to standard error.
    call logger%status('started ' // APP_NAME)

    rc = init(signal)
    if (dm_is_error(rc)) call shutdown(rc)

    rc = run(app, signal)
    call shutdown(rc)
contains
    integer function init(signal) result(rc)
        !! Initialises program.
        type(posix_signal_type), intent(out) :: signal !! Self-pipe.

        ! Create self-pipe and register signal handler.
        signal_block: block
            rc = dm_posix_signal_create(signal);                            if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGINT,  signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGQUIT, signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGABRT, signal_callback); if (dm_is_error(rc)) exit signal_block
            rc = dm_posix_signal_register(SIGNAL_SIGTERM, signal_callback); if (dm_is_error(rc)) exit signal_block
        end block signal_block

        if (dm_is_error(rc)) then
            call logger%error('failed to initialize signal handler', error=rc)
            return
        end if

        call logger%debug('initialized signal handler')
    end function init

    integer function output_observ(observ, type) result(rc)
        !! Outputs observation to file or _stdout_ if `type` is not
        !! `OUTPUT_NONE`.
        type(observ_type), intent(inout) :: observ !! Observation type.
        integer,           intent(in)    :: type   !! Output I/O type.

        integer :: stat, unit

        rc = E_NONE

        select case (type)
            case (OUTPUT_NONE)
                ! No output.
                return

            case (OUTPUT_STDOUT)
                ! Print to standard output.
                rc = write_observ(observ, unit=stdout, format=app%format)

                if (dm_is_error(rc)) then
                    call logger%error('failed to output observation', error=rc)
                    return
                end if

            case (OUTPUT_FILE)
                ! Write to file.
                rc = E_IO

                open (action='write', file=trim(app%output), iostat=stat, &
                      newunit=unit, position='append', status='unknown')

                if (stat /= 0) then
                    call logger%error('failed to open file ' // app%output, error=rc)
                    return
                end if

                rc = write_observ(observ, unit=unit, format=app%format)

                if (dm_is_error(rc)) then
                    call logger%error('failed to write observation to file ' // app%output, error=rc)
                end if

                close (unit)
        end select
    end function output_observ

    integer function read_observ(observ, node_id, sensor_id, source, debug) result(rc)
        !! Reads observation from file. By default, sends log messages to
        !! logger.
        !!
        !! The function returns the following error codes:
        !!
        !! * `E_IO` if reading from file failed.
        !! * `E_NOT_FOUND` if file could not be found.
        !!
        type(observ_type), intent(inout) :: observ    !! Observation to read.
        character(*),      intent(in)    :: node_id   !! Node id of observation.
        character(*),      intent(in)    :: sensor_id !! Sensor id of observation.
        character(*),      intent(in)    :: source    !! Source of observation.
        logical,           intent(in)    :: debug     !! Output debug messages.

        character(OBSERV_RESPONSE_LEN) :: raw
        integer                        :: i, msec, sec, unit, stat

        rc = E_EMPTY

        ! Initialise observation.
        call dm_observ_set(observ, id=dm_uuid4(), node_id=node_id, sensor_id=sensor_id, timestamp=dm_time_now(), source=source)
        if (debug) call logger%debug('started observation ' // observ%name, observ=observ)

        ! Return if observation is disabled.
        if (dm_observ_is_disabled(observ)) then
            if (debug) call logger%debug('observation ' // trim(observ%name) // ' is disabled', observ=observ)
            return
        end if

        ! Prepare observation.
        call dm_observ_set_response_error(observ, E_INCOMPLETE)

        ! Return if file path passed as observation request does not exist.
        if (.not. dm_file_exists(observ%request)) then
            rc = E_NOT_FOUND
            call logger%error('file ' // trim(observ%request) // ' not found', observ=observ, error=rc)
            return
        end if

        ! Try to open file for reading.
        io_block: block
            rc = E_IO
            open (action='read', file=trim(observ%request), iostat=stat, newunit=unit)

            if (stat /= 0) then
                call logger%error('failed to open file ' // trim(observ%request), observ=observ, error=rc)
                exit io_block
            end if

            ! Update time stamp.
            call dm_observ_set(observ, timestamp=dm_time_now())

            ! Read until the response pattern matches or end is reached.
            rc = E_EOF

            read_loop: do
                ! Read single line from unit.
                read (unit, '(a)', iostat=stat) raw
                if (is_iostat_end(stat)) exit read_loop

                rc = E_READ
                if (stat /= 0) exit read_loop

                ! Escape non-printable characters.
                observ%response = dm_ascii_escape(raw)

                ! Look for regular expression pattern.
                if (.not. dm_observ_has_pattern(observ)) then
                    rc = E_EMPTY
                    if (debug) call logger%debug('no regular expression pattern in observation ' // observ%name, observ=observ, error=rc)
                    exit read_loop
                end if

                ! Try to extract the response values.
                rc = dm_regex_observ(observ)

                if (dm_is_error(rc)) then
                    if (debug) call logger%debug('response in observation ' // trim(observ%name) // ' does not match pattern', observ=observ, error=rc)
                    cycle read_loop
                end if

                ! Look for response errors.
                do i = 1, observ%nresponses
                    associate (response => observ%responses(i))
                        if (dm_is_error(response%error)) then
                            call logger%warning('failed to extract response ' // trim(response%name) // ' in observation ' // observ%name, observ=observ, error=response%error)
                        end if
                    end associate
                end do

                exit read_loop
            end do read_loop
        end block io_block

        close (unit)
        if (dm_is_error(rc)) call logger%error('failed to read from file ' // observ%request, observ=observ, error=rc)
        if (debug) call logger%debug('finished observation ' // observ%name, observ=observ)

        msec = max(0, observ%delay)
        sec  = dm_msec_to_sec(msec)
        if (msec == 0) return
        if (debug) call logger%debug('next observation in ' // dm_itoa(msec) // ' sec', observ=observ)
        call dm_posix_msleep(msec)
    end function read_observ

    integer function run(app, signal) result(rc)
        !! Performs jobs in job list.
        type(app_type),          intent(inout) :: app    !! App type.
        type(posix_signal_type), intent(inout) :: signal !! Self-pipe.

        integer :: msec, sec
        integer :: next, njobs, number
        logical :: debug

        type(job_type)    :: job
        type(observ_type) :: observ

        debug = (app%debug .or. app%verbose)

        ! Run until no jobs are left.
        main_loop: do
            ! Poll for signals.
            rc = dm_posix_signal_poll(signal, timeout=0)

            if (rc == E_INTERRUPT) cycle main_loop

            if (dm_is_error(rc)) then
                call logger%error('failed to poll signal pipe', error=rc)
                exit main_loop
            end if

            if (dm_posix_signal_should_terminate(signal, number)) then
                call logger%debug('exit on signal ' // dm_posix_signal_name(number))
                exit main_loop
            end if

            ! Get next job as deep copy.
            njobs = dm_job_list_count(app%jobs)

            if (njobs == 0) then
                call logger%debug('no jobs left')
                exit main_loop
            end if

            if (debug) call logger%debug(dm_itoa(njobs) // dm_btoa((njobs == 1), ' job', ' jobs') // ' left in job queue')

            rc = dm_job_list_next(app%jobs, job)

            if (dm_is_error(rc)) then
                call logger%error('failed to fetch next job', error=rc)
                cycle main_loop
            end if

            if (dm_job_count(job) == 0) then
                call logger%debug('observation group of job is empty', error=E_EMPTY)
                cycle main_loop
            end if

            if (debug) call logger%debug('started job of observation group ' // dm_job_group_id(job))
            next = 0

            ! Run until no observations are left.
            observ_loop: do
                ! Get next observation of job group.
                rc = dm_job_next(job, next, observ)
                if (dm_is_error(rc)) exit observ_loop

                if (debug) call logger%debug('started observation ' // trim(observ%name) // ' of sensor ' // app%sensor_id, observ=observ)

                ! Read observation from file system.
                rc = read_observ(observ, app%node_id, app%sensor_id, app%name, debug)
                call dm_observ_set(observ, error=rc)

                ! Forward observation via message queue.
                rc = dm_posix_mqueue_forward(observ, name=app%name, blocking=APP_MQ_BLOCKING)

                ! Output observation.
                rc = output_observ(observ, app%output_type)
                if (debug) call logger%debug('finished observation ' // trim(observ%name) // ' of sensor ' // app%sensor_id, observ=observ)
            end do observ_loop

            if (debug) call logger%debug('finished job of observation group ' // dm_job_group_id(job))

            ! Wait delay time of the job if set (absolute).
            msec = max(0, job%delay)
            sec  = dm_msec_to_sec(sec)

            if (msec == 0) cycle
            if (debug) call logger%debug('next job in ' // dm_itoa(sec) // ' sec')
            call dm_posix_msleep(msec)
        end do main_loop
    end function run

    subroutine shutdown(error)
        !! Cleans up and stops program.
        integer, intent(in) :: error !! DMPACK error code.

        call dm_posix_signal_destroy(signal)
        call logger%status('stopped ' // APP_NAME, error=error)
        call dm_stop(merge(STOP_FAILURE, STOP_SUCCESS, dm_is_error(error)))
    end subroutine shutdown

    integer function write_observ(observ, unit, format) result(rc)
        !! Writes observation to file unit, in CSV or JSON Lines format.
        type(observ_type), intent(inout) :: observ !! Observation to write.
        integer,           intent(in)    :: unit   !! File unit.
        integer,           intent(in)    :: format !! Output format (`FORMAT_CSV`, `FORMAT_JSONL`).

        select case (format)
            case (FORMAT_CSV);   rc = dm_csv_write(observ, unit=unit, header=.false., separator=APP_CSV_SEPARATOR)
            case (FORMAT_JSONL); rc = dm_json_write(observ, unit=unit)
            case default;        rc = E_INVALID
        end select
    end function write_observ

    ! **************************************************************************
    ! COMMAND-LINE ARGUMENTS AND CONFIGURATION FILE.
    ! **************************************************************************
    integer function read_args(app) result(rc)
        !! Reads command-line arguments and settings from configuration file.
        type(app_type), intent(out) :: app !! App type.

        type(arg_parser_class) :: parser

        call parser%add('name',    short='n', type=ARG_TYPE_ID)      ! -n, --name <id>
        call parser%add('config',  short='c', type=ARG_TYPE_FILE, required=.true.) ! -c, --config <path>
        call parser%add('logger',  short='l', type=ARG_TYPE_ID)      ! -l, --logger <id>
        call parser%add('node',    short='N', type=ARG_TYPE_ID)      ! -N, --node <id>
        call parser%add('sensor',  short='S', type=ARG_TYPE_ID)      ! -S, --sensor <id>
        call parser%add('output',  short='o', type=ARG_TYPE_FILE)    ! -o, --output <path>
        call parser%add('format',  short='f', type=ARG_TYPE_STRING)  ! -f, --format <string>
        call parser%add('debug',   short='D', type=ARG_TYPE_LOGICAL) ! -D, --debug
        call parser%add('verbose', short='V', type=ARG_TYPE_LOGICAL) ! -V, --verbose

        ! Read all command-line arguments.
        rc = parser%read(version_callback)
        if (dm_is_error(rc)) return

        call parser%get('name',   app%name)
        call parser%get('config', app%config)

        ! Read configuration from file.
        rc = read_config(app)
        if (dm_is_error(rc)) return

        ! Get all other arguments.
        call parser%get('logger',  app%logger)
        call parser%get('node',    app%node_id)
        call parser%get('sensor',  app%sensor_id)
        call parser%get('output',  app%output)
        call parser%get('format',  app%format_name)
        call parser%get('debug',   app%debug)
        call parser%get('verbose', app%verbose)

        if (len_trim(app%output) > 0) then
            app%format = dm_format_from_name(app%format_name)

            if (trim(app%output) == '-') then
                app%output_type = OUTPUT_STDOUT
            else
                app%output_type = OUTPUT_FILE
            end if
        end if

        ! Validate options.
        rc = validate(app)
    end function read_args

    integer function read_config(app) result(rc)
        !! Reads configuration from (Lua) file.
        type(app_type), intent(inout) :: app !! App type.

        type(config_class) :: config

        rc = E_NONE
        if (len_trim(app%config) == 0) return

        rc = config%open(app%config, app%name)

        if (dm_is_ok(rc)) then
            call config%get('format',  app%format_name)
            call config%get('logger',  app%logger)
            call config%get('node',    app%node_id)
            call config%get('sensor',  app%sensor_id)
            call config%get('output',  app%output)
            call config%get('debug',   app%debug)
            call config%get('verbose', app%verbose)
            call config%get('jobs',    app%jobs, error=rc)
        end if

        call config%close()
    end function read_config

    integer function validate(app) result(rc)
        !! Validates options and prints error messages.
        type(app_type), intent(in) :: app !! App type.

        rc = E_INVALID

        if (.not. dm_id_is_valid(app%name)) then
            call dm_error_out(rc, 'invalid name')
            return
        end if

        if (.not. dm_id_is_valid(app%node_id)) then
            call dm_error_out(rc, 'invalid or missing node id')
            return
        end if

        if (.not. dm_id_is_valid(app%sensor_id)) then
            call dm_error_out(rc, 'invalid or missing sensor id')
            return
        end if

        if (len_trim(app%logger) > 0 .and. .not. dm_id_is_valid(app%logger)) then
            call dm_error_out(rc, 'invalid logger')
            return
        end if

        if (len_trim(app%output) > 0 .and. (app%format /= FORMAT_CSV .and. app%format /= FORMAT_JSONL)) then
            call dm_error_out(rc, 'invalid or missing output format')
            return
        end if

        rc = E_EMPTY

        if (dm_job_list_count(app%jobs) == 0) then
            call dm_error_out(rc, 'no enabled jobs')
            return
        end if

        rc = E_NONE
    end function validate

    ! **************************************************************************
    ! CALLBACKS.
    ! **************************************************************************
    subroutine signal_callback(number) bind(c)
        integer(c_int), intent(in), value :: number

        call dm_posix_signal_write(signal, number)
    end subroutine signal_callback

    subroutine version_callback()
        call dm_version_out(APP_NAME, APP_MAJOR, APP_MINOR, APP_PATCH)
        print '(a)', dm_lua_version(.true.)
    end subroutine version_callback
end program dmfs
